Src/Newtonsoft.Json/JsonWriter.Async.cs

```diff
@@ -1680,114 +1680,122 @@ protected Task SetWriteStateAsync(JsonToken token, object value, CancellationTok
+                        return writer.WriteValueAsync((string)value, cancellationToken);
+                    case PrimitiveTypeCode.Bytes:
+                        return writer.WriteValueAsync((byte[])value, cancellationToken);
 #if HAVE_DB_NULL_TYPE_CODE
-                case PrimitiveTypeCode.DBNull:
-                    return writer.WriteNullAsync(cancellationToken);
+                    case PrimitiveTypeCode.DBNull:
+                        return writer.WriteNullAsync(cancellationToken);
 #endif
-                default:
+                    default:
 #if HAVE_ICONVERTIBLE
-                    if (value is IConvertible convertable)
-                    {
-                        // the value is a non-standard IConvertible
-                        // convert to the underlying value and retry
-                        TypeInformation typeInformation = ConvertUtils.GetTypeInformation(convertable);
-
-                        // if convertible has an underlying typecode of Object then attempt to convert it to a string
-                        PrimitiveTypeCode resolvedTypeCode = typeInformation.TypeCode == PrimitiveTypeCode.Object ? PrimitiveTypeCode.String : typeInformation.TypeCode;
-                        Type resolvedType = typeInformation.TypeCode == PrimitiveTypeCode.Object ? typeof(string) : typeInformation.Type;
+                        if (value is IConvertible convertible)
+                        {
+                            // the value is a non-standard IConvertible
+                            // convert to the underlying value and retry
+                            TypeInformation typeInformation = ConvertUtils.GetTypeInformation(convertible);
+
+                            // if convertible has an underlying typecode of Object then attempt to convert it to a string
+                            typeCode = typeInformation.TypeCode == PrimitiveTypeCode.Object ? PrimitiveTypeCode.String : typeInformation.TypeCode;
+                            Type resolvedType = typeInformation.TypeCode == PrimitiveTypeCode.Object ? typeof(string) : typeInformation.Type;
+                            value = convertible.ToType(resolvedType, CultureInfo.InvariantCulture);
+                            continue;
+                        }
+#endif
 
-                        object convertedValue = convertable.ToType(resolvedType, CultureInfo.InvariantCulture);
+                        // write an unknown null value, fix https://github.com/JamesNK/Newtonsoft.Json/issues/1460
+                        if (value == null)
+                        {
+                            return writer.WriteNullAsync(cancellationToken);
+                        }
 
-                        return WriteValueAsync(writer, resolvedTypeCode, convertedValue, cancellationToken);
-                    }
-#endif
-                    throw CreateUnsupportedTypeException(writer, value);
+                        throw CreateUnsupportedTypeException(writer, value);
```

Comment:
Should this be throwing in Task returning, non-async method? Should the ex be returned via `Task.FromException`?

Reply:
To my mind if one can tell straight away that it's impossible to create a task that will be successful it's better to throw immediately outside of the task mechanism (comparable to how `default(IEnumerable<int>).Select(i => i)` throws immediately while `actualSequence.Select(funcThatThrows)` doesn't throw until enumeration), but I don't go so far as to make sure it happens (doing validation in a non `async` method and then calling into an `async`). I probably should though.

It would be good to be consistent, one way or another.

Reply:
Ok, that makes sense. I might go through and audit the thrown exceptions in Task methods that aren't async.